### PR TITLE
handle promise rejections as well

### PR DESCRIPTION
**Pull request by Alahel**
* handle promises rejections in the chain

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
   slice?: [number, number],
 };
 
-type Queue = { resolve: (val: any) => any, args: any[] }[];
+type QueueItem = { resolve: (val: any) => any, reject: (val: any) => any, args: any[] }
```
Comment (grainrigi) [inline src/index.ts]: Do not omit semicolon

File: src/index.ts [hunk 2]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
...
   if (typeof func !== 'function') throw new Error('argument is not function.');
   const contexts: { args: any[], queue: Queue }[] = [];
   const slicer: (args: any[]) => any[] = slice ? v => v.slice(slice[0], slice[1]) : v => v;
   return function (this: any, ...args: any[]) {
     // eslint-disable-next-line no-async-promise-executor
-    return new Promise(async (resolve) => {
+    return new Promise(async (resolve, reject) => {
       // コンテキストを用意
-      let context = find(contexts, v => isEqual(slicer(args), slicer(v.args)));
+      let context = find(contexts, (v: any) => isEqual(slicer(args), slicer(v.args)));
```
Comment (grainrigi) [inline src/index.ts]: Unnecessarry type annotation since element type can be inferred

File: src/index.ts [hunk 3]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
...
         context = { args, queue: [] };
         contexts.push(context);
       }
 
-      if (running) return context.queue.push({ resolve, args });
-      let result = await func.call(this, ...args);
-      resolve(result);
+      if (running) return context.queue.push({ resolve, reject, args });
+      let result: any
+      let errorResult: any
```
Comment (grainrigi) [inline src/index.ts]: Do not omit semicolon

File: src/index.ts [hunk 4]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
...
+      try {
+        result = await func.call(this, ...args);
+        resolve(result);
+      } catch (e) {
+        errorResult = e
+        reject(errorResult)
```
Comment (grainrigi) [inline src/index.ts]: Do not omit semicolon

File: src/index.ts [hunk 5]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
...
+      }
       if (context.queue.length > 0) {
         if (trailing) {
           const { args } = last(context.queue);
-          result = await func.call(this, ...args);
+          try {
+            result = await func.call(this, ...args);
+          } catch (e) {
+            errorResult = e
```
Comment (grainrigi) [inline src/index.ts]: Do not omit semicolon

File: src/index.ts [hunk 6]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
...
+          }
         }
-        context.queue.forEach(({ resolve }) => resolve(result));
+        context.queue.forEach(({ resolve, reject }: QueueItem) => {
+          if (errorResult) return reject(errorResult)
+          return resolve(result)
```
Comment (grainrigi) [inline src/index.ts]: Do not omit semicolon

File: src/index.ts [hunk 7]
```diff
@@ -7,37 +7,52 @@ export type AsyncSingletonOption = {
...
+        });
         context.queue = [];
       }
 
       // コンテキストを削除
-      remove(contexts, v => isEqual(slicer(args), slicer(v.args)));
+      remove(contexts, (v: any) => isEqual(slicer(args), slicer(v.args)));
```
Comment (grainrigi) [inline src/index.ts]: Unnecessarry type annotation

**Review by grainrigi — CHANGES_REQUESTED** [review]
Thank you for your contribution.
Please check the reviewes before I can merge your request.

**Comment by Alahel** [comment]
> Thank you for your contribution.
> Please check the reviewes before I can merge your request.

Thanks for the review :+1:

**Comment by grainrigi** [comment]
Everything looks good to me.
Thanks!

**Pull request merged**